airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_with_global.py

```diff
@@ -0,0 +1,179 @@
+
+    This cursor handles partitioned streams by maintaining individual state per partition using `PerPartitionCursor`. If the number of partitions exceeds a defined limit, it switches to a global cursor (`GlobalSubstreamCursor`) to manage state more efficiently.
+
+    **Overview**
+
+    - **Partition-Based State**: Initially manages state per partition to ensure accurate processing of each partition's data.
+    - **Global Fallback**: Switches to a global cursor when the partition limit is exceeded to handle state management more effectively.
+
+    **Switching Logic**
+
+    - Monitors the number of partitions.
+    - If `PerPartitionCursor.limit_reached()` returns `True`, sets `_use_global_cursor` to `True`, activating the global cursor.
+
+    **Active Cursor Selection**
+
+    - Uses the `_get_active_cursor()` helper method to select the active cursor based on the `_use_global_cursor` flag.
+    - This simplifies the logic and ensures consistent cursor usage across methods.
+
+    **State Structure Example**
+
+    ```json
+    {
+        "states": [
+            {
+                "partition": {"partition_key": "partition_1"},
+                "cursor": {"cursor_field": "2021-01-15"}
+            },
+            {
+                "partition": {"partition_key": "partition_2"},
+                "cursor": {"cursor_field": "2021-02-14"}
+            }
+        ],
+        "state": {
+            "cursor_field": "2021-02-15"
+        },
+        "use_global_cursor": false
+    }
+    ```
+
+    In this example, the cursor is using partition-based state management (`"use_global_cursor": false`), maintaining separate cursor states for each partition.
+
+    **Usage Scenario**
+
+    Suitable for streams where the number of partitions may vary significantly, requiring dynamic switching between per-partition and global state management to ensure data consistency and efficient synchronization.
+    """
+
+    def __init__(self, cursor_factory: CursorFactory, partition_router: PartitionRouter, stream_cursor: DatetimeBasedCursor):
+        self._partition_router = partition_router
+        self._per_partition_cursor = PerPartitionCursor(cursor_factory, partition_router)
+        self._global_cursor = GlobalSubstreamCursor(stream_cursor, partition_router)
+        self._use_global_cursor = False
+
+    def _get_active_cursor(self) -> Union[PerPartitionCursor, GlobalSubstreamCursor]:
+        return self._global_cursor if self._use_global_cursor else self._per_partition_cursor
+
+    def stream_slices(self) -> Iterable[StreamSlice]:
+        self._global_cursor.start_slices_generation()
+
+        # Iterate through partitions and process slices
+        for partition, is_last_partition in iterate_with_last_flag(self._partition_router.stream_slices()):
+            # Generate slices for the current cursor and handle the last slice using the flag
+            if partition is None:
```

Comment:
Why do we need this if? In which case can this happen?  Why do we simply `continue`? Should we maybe log? 

The same questions apply for `if slice is None:` below. Depending on the answers, it feels like it might be interesting to have unit tests about those

Reply:
I refactored the `iterate_with_last_flag` to avoid skipping None.